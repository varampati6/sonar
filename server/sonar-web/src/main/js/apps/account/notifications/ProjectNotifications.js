/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import NotificationsList from './NotificationsList';
import Organization from '../../../components/shared/Organization';
import { translate } from '../../../helpers/l10n';
import {
  getProjectNotifications,
  getNotificationChannels,
  getNotificationPerProjectTypes
} from '../../../store/rootReducer';
import type {
  Notification,
  NotificationsState,
  ChannelsState,
  TypesState
} from '../../../store/notifications/duck';
import { addNotification, removeNotification } from './actions';
import { getProjectUrl } from '../../../helpers/urls';

class ProjectNotifications extends React.PureComponent {
  props: {
    project: {
      key: string,
      name: string,
      organization: string
    },
    notifications: NotificationsState,
    channels: ChannelsState,
    types: TypesState,
    addNotification: (n: Notification) => void,
    removeNotification: (n: Notification) => void
  };

  handleAddNotification({ channel, type }) {
    this.props.addNotification({
      channel,
      type,
      project: this.props.project.key,
      projectName: this.props.project.name,
      organization: this.props.project.organization
    });
  }

  handleRemoveNotification({ channel, type }) {
    this.props.removeNotification({
      channel,
      type,
      project: this.props.project.key
    });
  }

  render() {
    const { project, channels } = this.props;

    return (
      <table key={project.key} className="form big-spacer-bottom">
        <thead>
          <tr>
            <th>
              <span className="text-normal">
                <Organization organizationKey={project.organization} />
              </span>
              <h4 className="display-inline-block">
                <Link to={getProjectUrl(project.key)}>{project.name}</Link>
              </h4>
            </th>
            {channels.map(channel => (
              <th key={channel} className="text-center">
                <h4>{translate('notification.channel', channel)}</h4>
              </th>
            ))}
          </tr>
        </thead>
        <NotificationsList
          notifications={this.props.notifications}
          channels={this.props.channels}
          types={this.props.types}
          checkboxId={(d, c) => `project-notification-${project.key}-${d}-${c}`}
          onAdd={n => this.handleAddNotification(n)}
          onRemove={n => this.handleRemoveNotification(n)}
        />
      </table>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  notifications: getProjectNotifications(state, ownProps.project.key),
  channels: getNotificationChannels(state),
  types: getNotificationPerProjectTypes(state)
});

const mapDispatchToProps = { addNotification, removeNotification };

export default connect(mapStateToProps, mapDispatchToProps)(ProjectNotifications);

export const UnconnectedProjectNotifications = ProjectNotifications;
