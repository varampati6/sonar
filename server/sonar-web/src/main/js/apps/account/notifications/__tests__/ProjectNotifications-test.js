/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import { UnconnectedProjectNotifications } from '../ProjectNotifications';
import NotificationsList from '../NotificationsList';

const channels = ['channel1', 'channel2'];
const types = ['type1', 'type2'];
const notifications = [
  { channel: 'channel1', type: 'type1' },
  { channel: 'channel1', type: 'type2' },
  { channel: 'channel2', type: 'type2' }
];

it('should match snapshot', () => {
  expect(
    shallow(
      <UnconnectedProjectNotifications
        project={{ key: 'foo', name: 'Foo' }}
        notifications={notifications}
        channels={channels}
        types={types}
        addNotification={jest.fn()}
        removeNotification={jest.fn()}
      />
    )
  ).toMatchSnapshot();
});

it('should call `addNotification` and `removeNotification`', () => {
  const addNotification = jest.fn();
  const removeNotification = jest.fn();
  const wrapper = shallow(
    <UnconnectedProjectNotifications
      project={{ key: 'foo', name: 'Foo' }}
      notifications={notifications}
      channels={channels}
      types={types}
      addNotification={addNotification}
      removeNotification={removeNotification}
    />
  );
  const notificationsList = wrapper.find(NotificationsList);

  notificationsList.prop('onAdd')({ channel: 'channel2', type: 'type1' });
  expect(addNotification).toHaveBeenCalledWith({
    channel: 'channel2',
    type: 'type1',
    project: 'foo',
    projectName: 'Foo'
  });

  jest.resetAllMocks();

  notificationsList.prop('onRemove')({ channel: 'channel1', type: 'type1' });
  expect(removeNotification).toHaveBeenCalledWith({
    channel: 'channel1',
    type: 'type1',
    project: 'foo'
  });
});
