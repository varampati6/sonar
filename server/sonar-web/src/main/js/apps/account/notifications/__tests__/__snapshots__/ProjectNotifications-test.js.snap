// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<table
  className="form big-spacer-bottom"
>
  <thead>
    <tr>
      <th>
        <span
          className="text-normal"
        >
          <Connect(Organization) />
        </span>
        <h4
          className="display-inline-block"
        >
          <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/dashboard",
                "query": Object {
                  "id": "foo",
                },
              }
            }
          >
            Foo
          </Link>
        </h4>
      </th>
      <th
        className="text-center"
      >
        <h4>
          notification.channel.channel1
        </h4>
      </th>
      <th
        className="text-center"
      >
        <h4>
          notification.channel.channel2
        </h4>
      </th>
    </tr>
  </thead>
  <NotificationsList
    channels={
      Array [
        "channel1",
        "channel2",
      ]
    }
    checkboxId={[Function]}
    notifications={
      Array [
        Object {
          "channel": "channel1",
          "type": "type1",
        },
        Object {
          "channel": "channel1",
          "type": "type2",
        },
        Object {
          "channel": "channel2",
          "type": "type2",
        },
      ]
    }
    onAdd={[Function]}
    onRemove={[Function]}
    types={
      Array [
        "type1",
        "type2",
      ]
    }
  />
</table>
`;
